manta-util/src/collections/btree_map.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! An ordered map based on a B-Tree.
+
+use core::borrow::Borrow;
+
+#[doc(inline)]
+pub use alloc::collections::btree_map::*;
+
+/// Error Message for the [`pop_last`] Function
+const POP_LAST_ERROR_MESSAGE: &str =
+    "Querying the last element of the map must guarantee that `BTreeMap::remove` always succeeds.";
+
+/// Pops the last element in the key-ordering of `map`.
+///
+/// # Limitation
+///
+/// Until <https://github.com/rust-lang/rust/issues/62924> is stabilized this is an equivalent
+/// way to pop the last key-value pair out of a [`BTreeMap`] but requires cloning the key.
+#[inline]
+pub fn pop_last<K, V>(map: &mut BTreeMap<K, V>) -> Option<(K, V)>
+where
+    K: Clone + Ord,
+{
+    let key = map.keys().last()?.clone();
+    match map.remove(&key) {
+        Some(value) => Some((key, value)),
+        _ => unreachable!("{}", POP_LAST_ERROR_MESSAGE),
```

Comment:
Why panic instead of returning `None`?

Reply:
Because we are guaranteed by the call on line `39` that this state is unreachable.